### PR TITLE
kernel: handle thread self-aborts on idle thread

### LINKED ISSUE #26486: possible SMP race with k_thread_join()
**Describe the bug**
From code inspection I think I have found a potential race condition in k_thread_join().

The specific scenario:

1. Thread A is running
2. Thread B blocks with `k_thread_join(A, K_FOREVER);`
3. Thread A self-exits via `k_thread_abort(k_current_get())` either via explicit call or by returning from its entry function
4. In the code path for `k_thread_abort()`, we enter `z_thread_single_abort()`. While holding the `sched_spinlock`, we wake up waiters, in this case thread B. The `sched_spinlock` is released.
5. Two things then happen simultaneously:
* On the current CPU, thread A finishes its aborting process. We return from `z_thread_single_abort()` and then call `z_swap_unlocked()` to finally context switch out. This happens quickly but isn't instantaneous.
* On another CPU, as soon as `sched_spinlock` was released in step 4, thread B wakes up. Since thread A has joined, thread B tries to launch a new thread re-using thread A's `k_thread` object.

The race is step 5. If thread B manages to instantiate the new thread using thread A's object, before thread A finishes its aborting process, we will end up with the same thread object (and the stack too if it was re-used) being active on two different CPUs causing data corruption.

This only affects SMP. In UP, if Thread A gets preempted after returning from `z_thread_single_abort()` it is never scheduled again and there is no race. This issue requires physical concurrency AFAICT.

**To Reproduce**
I don't have test code to demonstrate this reliably but it would explain some very intermittent crashes I see in tests that re-use thread objects after a join where both CPUs are crashing with the same thread object as current.

**Expected behavior**
We need to close this race somehow. Thread B needs to not wake up until thread A's 	`k_thread` and thread stack objects are truly unused.

It would of course be simplest to somehow have thread A keep holding `sched_spinlock` until it completely finishes aborting, but this is not feasible.

A promising alternative is to somehow bounce context off of Thread A to complete the aborting process; we could enter a special exception state (sort of like what `irq_offload()` does although this API was never intended for anything but test usage), or wake up and raise the priority of the idle thread and have the idle thread perform the logic in `z_thread_single_abort()`.

I like the idea of using the idle thread since we already have one per CPU and we could avoid having thread abort tasks being queued up.

In general, we never want `z_thread_single_abort()` to run in the context of the thread being aborted.

**Impact**
This is concerning, but to be clear this only affects SMP and it requires a race to re-use a thread object, something that I expect happens more often in tests than in a real application.

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Fixes races where threads on another CPU are joining the
exiting thread, since it could still be running when
the joiners wake up on a different CPU.

Fixes problems where the thread object is still being
used by the kernel when the fn_abort() function is called,
preventing the thread object from being recycled or
freed back to a slab pool.

Fixes a race where a thread is aborted from one CPU while
it self-aborts on another CPU, that was currently worked
around with a busy-wait.

Precedent for doing this comes from FreeRTOS, which also
performs final thread cleanup in the idle thread.

Add some extra comments and log messages to the meta-irq
test case.

Some logic in z_thread_single_abort() rearranged such that
when we release sched_spinlock, the thread object pointer
is never dereferenced by the kernel again; join waiters
or fn_abort() logic may free it immediately.

An assertion added to z_thread_single_abort() to ensure
it never gets called with thread == _current outside of an ISR.

Some logic has been added to ensure z_thread_single_abort()
tasks don't run more than once.

Add a test proving the idle thread stack size is large
enough.

Added another patch to detect via assertion whether a
thread objects are being re-used when they are still
active. horrible and difficult to comprehend things can
happen if a thread object is already being used by the
kernel and k_thread_create() is called on it.

Fixes: #26486
Related to: #23063 #23062

### CODE REVIEW

**Comment by andrewboie** [comment]
debugging some issues in tests that abort out of a fatal error, due to inconsistencies in how various arches report `is_in_isr()`...

File: kernel/thread_abort.c [hunk 1]
```diff
@@ -21,25 +21,48 @@
...
+	 */
+	int key;
+	struct _cpu *cpu;
+
+	/* Lock local IRQs to prevent us from migrating to another CPU
+	 * while we set this up
+	 */
+	key = arch_irq_lock();
+	cpu = _current_cpu;
+	cpu->pending_abort = _current;
```
Comment (nashif) [inline kernel/thread_abort.c:40]: shouldn't this be a list to handle multiple threads self-aborting before we are able to switch to idle to cleanup?

Comment (andrewboie) [reply kernel/thread_abort.c:40]: can you give some detail how that could happen?

I think if a thread self-aborts, the next thing that gets scheduled is necessarily the idle thread on the same CPU that the thread aborted on. I don't see how multiple could queue up since it should impossible for another thread to run in between on this CPU.

Comment (nashif) [reply kernel/thread_abort.c:40]: Just noticed the change in next_up() which will put the idle thread next if there is something to abort, so you are correct.

Before:

```
1969-12-31 19:00:00.005629 (+0.000148 s): thread_pending: thread_b
1969-12-31 19:00:00.005778 (+0.000150 s): thread_switched_out: thread_b
1969-12-31 19:00:00.005829 (+0.000050 s): thread_switched_in: thread_c
1969-12-31 19:00:00.005887 (+0.000058 s): thread_abort: thread_c
1969-12-31 19:00:00.005937 (+0.000050 s): thread_switched_out: thread_c
1969-12-31 19:00:00.005985 (+0.000048 s): thread_switched_in: thread_d
1969-12-31 19:00:00.006050 (+0.000065 s): thread_abort: thread_d
1969-12-31 19:00:00.006110 (+0.000061 s): thread_switched_out: thread_d
1969-12-31 19:00:00.006159 (+0.000049 s): thread_switched_in: idle 00
1969-12-31 19:00:00.006295 (+0.000135 s): idle
```
2 threads are abort, then we go into idle..

Now
```
1969-12-31 19:00:00.005027 (+0.000041 s): thread_switched_in: thread_c
1969-12-31 19:00:00.005075 (+0.000048 s): thread_suspend: thread_c
1969-12-31 19:00:00.005116 (+0.000041 s): thread_switched_out: thread_c
1969-12-31 19:00:00.005156 (+0.000040 s): thread_switched_in: idle 00
1969-12-31 19:00:00.005222 (+0.000066 s): thread_abort: thread_c
1969-12-31 19:00:00.005273 (+0.000051 s): thread_switched_out: idle 00
1969-12-31 19:00:00.005313 (+0.000040 s): thread_switched_in: thread_d
1969-12-31 19:00:00.005364 (+0.000051 s): thread_suspend: thread_d
1969-12-31 19:00:00.005405 (+0.000041 s): thread_switched_out: thread_d
1969-12-31 19:00:00.005444 (+0.000039 s): thread_switched_in: idle 00
1969-12-31 19:00:00.005485 (+0.000041 s): thread_abort: thread_d
1969-12-31 19:00:00.005536 (+0.000051 s): thread_switched_out: idle 00
1969-12-31 19:00:00.005577 (+0.000041 s): thread_switched_in: idle 00
```

File: kernel/sched.c [hunk 2]
```diff
@@ -501,7 +510,6 @@ void z_thread_single_abort(struct k_thread *thread)
 						 thread);
 				z_mark_thread_as_not_queued(thread);
 			}
-			update_cache(thread == _current);
```
Comment (andrewboie) [inline kernel/sched.c:504]: oops, this needs to remain

**Comment by andrewboie** [comment]
Somehow, this has introduced an order-of-magnitude regression on how fast tests/lib/heap completes on qemu_xtensa.

I have no idea why at this time, but is is easy to reproduce.

**Comment by andrewboie** [comment]
This xtensa issue is very mysterious. I don't know how to gather gprof data on an emulated target (is this possible?), so I don't have a good source of how much time the CPU is spending inside various functions.

Just randomly ctrl-c in GDB while the test is running reveals that we're spending our time in the context of `sys_heap_stress` and the IP is typically within one of the `chunk_` inline functions, most often in `chunk_field`. A typical backtrace:

```
(gdb) bt
#0  chunk_field (h=0x60026ee0 <heapmem>, c=<optimized out>, f=SIZE_AND_USED)
    at /home/apboie/projects/zephyr/zephyr/lib/os/heap.h:94
#1  0x600023dc in chunk_size (h=0x60026ee0 <heapmem>, c=5077)
    at /home/apboie/projects/zephyr/zephyr/lib/os/heap.h:125
#2  0x600024fc in valid_chunk (h=0x60026ee0 <heapmem>, c=5077)
    at /home/apboie/projects/zephyr/zephyr/lib/os/heap-validate.c:37
#3  0x600025e9 in sys_heap_validate (heap=<optimized out>)
    at /home/apboie/projects/zephyr/zephyr/lib/os/heap-validate.c:84
#4  0x6000175a in testfree (arg=0x60005828 <ztest_thread_stack+1688>,
    p=0x600286d4 <heapmem+6132>) at ../src/main.c:117
#5  0x60002903 in sys_heap_stress (alloc=0x6000175c <testalloc>, free=0x600016c8 <testfree>,
    arg=0x60005828 <ztest_thread_stack+1688>, total_bytes=262144, op_count=4096,
    scratch_mem=0x60006ee0 <scratchmem>, scratch_bytes=scratch_bytes@entry=131072,
    target_percent=target_percent@entry=100,
    result=result@entry=0x60005810 <ztest_thread_stack+1664>)
    at /home/apboie/projects/zephyr/zephyr/lib/os/heap-validate.c:310
#6  0x600016bf in test_big_heap () at ../src/main.c:199
#7  0x600031c9 in run_test_functions (test=<optimized out>)
    at /home/apboie/projects/zephyr/zephyr/subsys/testsuite/ztest/src/ztest.c:191
#8  test_cb (a=<optimized out>, dummy2=<optimized out>, dummy=<optimized out>)
    at /home/apboie/projects/zephyr/zephyr/subsys/testsuite/ztest/src/ztest.c:321
#9  0x60001c12 in z_thread_entry (entry=0x600031ac <test_cb>,
    p1=0x60006cc0 <_lib_heap_test$4433+40>, p2=0x0, p3=0x0)
    at /home/apboie/projects/zephyr/zephyr/lib/os/thread_entry.c:29
```

I also occasionally will break and find myself within `_WindowOverflow8`. But this also happens in a vanilla tree, so all I can do is speculate whether there has been some subtle change that is causing these window overflow exceptions to happen much more often. I could be barking completely up the wrong tree, I don't know this arch well.

I am not sure how to gather data to investigate further; gprof output would be valuable, or perhaps a counter value for each exception occurrence.

This is surely very weird as the test is single threaded and isn't sleeping, so why these changes I made would cause this particular test on this particular arch to perform an order of magnitude worse has me baffled.

File: kernel/Kconfig [hunk 3]
```diff
@@ -156,9 +156,7 @@ config IDLE_STACK_SIZE
 	int "Size of stack for idle thread"
 	default 2048 if COVERAGE_GCOV
 	default 1024 if XTENSA
-	default 512 if RISCV
-	default 320 if ARC || (ARM && CPU_HAS_FPU)
-	default 256
+	default 512
```
Comment (ioannisg) [inline kernel/Kconfig:159]: Can we leave 256 for ARM boards with little ram, and in particular the v6-M architecture which only needs to stack the standard ESF?

Comment (andrewboie) [reply kernel/Kconfig:159]: it would depend on whether leaving at 256 would be enough to prevent a stack overflow of the idle thread, since it now has to occasionally call into `k_thread_single_abort()`. If you have any such boards, some data would be helpful (I unfortunately do not)

Comment (andrewboie) [reply kernel/Kconfig:159]: I expect those boards wouldn't have an MPU so I imagine the best way to validate would be to either enable CONFIG_STACK_SENTINEL or CONFIG_INIT_STACKS and do some analysis after some threads enter/exit. I'll add a test for this in this PR.

Comment (andrewboie) [reply kernel/Kconfig:159]: Turns out we have some boards. Will follow up.

Comment (ioannisg) [reply kernel/Kconfig:159]: thanks. There are a few nucleo boards or even nrf51 with 16kB of ram only.

Comment (andrewboie) [reply kernel/Kconfig:159]: I have good news, we can leave the stack sizes as-is.

**Comment by andrewboie** [comment]
Updated to leave idle stack sizes alone (either an older version of this needed this, or some instrumentation I had, either way it's not necessary in this version) and added an idle thread stack size test case.

**Comment by andrewboie** [comment]
> Updated to leave idle stack sizes alone (either an older version of this needed this, or some instrumentation I had, either way it's not necessary in this version) and added an idle thread stack size test case.

Ah, not quite true. The overflows I saw, however, are pinpointed to CONFIG_DYNAMIC_OBJECTS hooks to free them when references drop to zero. Adjusted kernel/Kconfig appropriately. I want to remove this magic for 2.5.

**Comment by andrewboie** [comment]
Added another patch to detect via assertion whether a
thread objects are being re-used when they are still
active. horrible and difficult to comprehend things can
happen if a thread object is already being used by the
kernel and k_thread_create() is called on it.

**Comment by andrewboie** [comment]
The thread re-use patch exposed an SMP issue on x86_64 in CI that is fixed by https://github.com/zephyrproject-rtos/zephyr/pull/28102

File: kernel/sched.c [hunk 4]
```diff
@@ -511,53 +534,62 @@ void z_thread_single_abort(struct k_thread *thread)
...
-		 * thread's stack and that the abort is known to be
-		 * complete.  Otherwise the thread still runs a bit
-		 * until it can swap, requiring a delay.
+		/* Wake everybody up who was trying to join with this thread.
+		 * A reschedule is invoked later by k_thread_abort().
 		 */
-		if (IS_ENABLED(CONFIG_SMP) && arch_is_in_isr()) {
-			mask |= _THREAD_ABORTED_IN_ISR;
+		while ((waiter = z_waitq_head(&thread->base.join_waiters)) !=
+		       NULL) {
```
Comment (andyross) [inline kernel/sched.c:542]: Note that wait queues are priority-ordered, so there's a performance cost to using it as the "list backend" for this purpose.  Longer term we might consider unioning a slist or something into the same node memory.

File: tests/kernel/sched/metairq/src/main.c [hunk 5]
```diff
@@ -59,7 +59,7 @@ void metairq_thread(void)
 	printk("give sem1\n");
 	k_sem_give(&coop_sem1);
 
-	printk("metairq end\n");
+	printk("metairq end, should switch back to co-op thread2\n");
```
Comment (andyross) [inline tests/kernel/sched/metairq/src/main.c:62]: Seems like the changes to this test are just extra verbosity and a new comment to explain the sequencing?  Could remove, or put in a separate commit.

File: tests/kernel/threads/thread_stack/src/main.c [hunk 6]
```diff
@@ -414,10 +414,76 @@ void test_stack_buffer(void)
...
+/**
+ * @brief Show that the idle thread stack size is correct
+ *
+ * The idle thread has to occasionally clean up self-exiting threads.
+ * Exercise this and show that we didn't overflow, reporting out stack
+ * usage.
+ *
+ * @ingroup kernel_memprotect_tests
+ */
+void test_idle_stack(void)
```
Comment (andyross) [inline tests/kernel/threads/thread_stack/src/main.c:442]: Add this new test case as a separate commit?

File: kernel/idle.c [hunk 7]
```diff
@@ -152,6 +157,24 @@ void idle(void *unused1, void *unused2, void *unused3)
...
+
+		if (to_abort) {
+			cpu->pending_abort = NULL;
+			arch_irq_unlock(key);
+
+			LOG_DBG("idle %p aborting thread %p",
+				_current, to_abort);
+
+			z_thread_single_abort(to_abort);
+			z_reschedule_unlocked();
```
Comment (andyross) [inline kernel/idle.c:174]: This bit needs a comment, I think.  Normally, the fact that the idle thread is running at all means by definition that no other higher priority threads are available.  So a reschedule here really should be a noop.

But it's not now, because the thread abort case means that the idle thread can also run as the "highest" priority too, in which case it can't ignore the reschedule because it would start spinning until the next interrupt.

That's subtle and surprising, so we should call it out here so no one comes by and tries to remove the reschedule to be helpful.

File: kernel/idle.c [hunk 8]
```diff
@@ -152,6 +157,24 @@ void idle(void *unused1, void *unused2, void *unused3)
...
+		struct k_thread *to_abort = cpu->pending_abort;
+
+		if (to_abort) {
+			cpu->pending_abort = NULL;
+			arch_irq_unlock(key);
+
+			LOG_DBG("idle %p aborting thread %p",
+				_current, to_abort);
+
+			z_thread_single_abort(to_abort);
```
Comment (andyross) [inline kernel/idle.c:173]: Doing this outside the look seems racy?  Maybe not a problem in practice because of the flag state, but does z_thread_single_abort() really need to be called with interrupts unmasked?

**Review by andyross — APPROVED** [review]
Some nitpicks.  Logic all looks correct to me.

File: kernel/idle.c (hunk 8 above)
Comment (andrewboie) [reply kernel/idle.c:173]: > Doing this outside the look seems racy?

I think this is OK. We've atomically checked and stashed the `cpu->pending_abort` value to a stack variable. If we get preempted before `z_thread_single_abort()` is called, and another thread aborts, `cpu->pending_abort` will be set and we'll get right back here. We'll then abort the first thread, and then repeat the loop iteration to get the second one. I'll add a comment explaining this.

>  does z_thread_single_abort() really need to be called with interrupts unmasked?

It doesn't. I thought it would be a little better for latency, especially if thread->fn_abort has a function attached.

File: kernel/idle.c (hunk 7 above)
Comment (andrewboie) [reply kernel/idle.c:174]: will do

File: kernel/sched.c (hunk 4 above)
Comment (andrewboie) [reply kernel/sched.c:542]: Agreed, I think we spoke about this a bit in the original `k_thread_join()` patch, I've filed an enhancement here: https://github.com/zephyrproject-rtos/zephyr/issues/28609

File: tests/kernel/sched/metairq/src/main.c (hunk 5 above)
Comment (andrewboie) [reply tests/kernel/sched/metairq/src/main.c:62]: agreed will split this out. I added this stuff because one version of this PR failed this test, and I had to spend some time trying to figure out how this test was supposed to behave.

File: tests/kernel/threads/thread_stack/src/main.c (hunk 6 above)
Comment (andrewboie) [reply tests/kernel/threads/thread_stack/src/main.c:442]: will do

**Review by tyrantlizard77 — COMMENTED** [review]
OK updated and rebased.

- test case changes split into their own commits
- added comments to idle code explaining some subtleties

**Comment by andrewboie** [comment]
With help from QA who had some real Xtensa boards, we found that the issue with the heap test is not reproducible on real xtensa hardware. On an Intel S1000 ASDP, the test ran in 3.7 seconds with and without this PR.

I added a patch to disable the heap test on qemu_xtensa, this is some weird QEMU issue.

**Review by nashif — APPROVED** [review]

File: arch/arm/core/aarch32/cortex_m/thread_abort.c [hunk 9]
```diff
@@ -43,10 +41,12 @@ void z_impl_k_thread_abort(k_tid_t thread)
 			 */
 			SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
 		} else {
-			z_swap_unlocked();
+			z_self_abort();
```
Comment (ioannisg) [inline arch/arm/core/aarch32/cortex_m/thread_abort.c:44]: @andrewboie do we need CODE_UNREACHABLE after this line, here?
Or at least a comment that indicates this, for ease of reading...

**Comment by ruuddw** [comment]
Looks ok to me, and few quick ARC tests all pass. Do have some concerns with changing something like this so late in the release cycle, as there always can be corner cases not foreseen or covered by the tests and new behavior not fully understood (xtensa qemu issue). Can it wait till next release?

**Comment by evgeniy-paltsev** [comment]
Hmm, I've tried to run tests on HSDK (ARC HW SMP platform) and I got all tests failed. I'll investigate it.

**Comment by evgeniy-paltsev** [comment]
Ok, it was issue with our verification and it's not related to this PR.

File: arch/arm/core/aarch32/cortex_m/thread_abort.c (hunk 9 above)
Comment (andrewboie) [reply arch/arm/core/aarch32/cortex_m/thread_abort.c:44]: @ioannisg `z_self_abort()` is `FUNC_NORETURN` but I added a `CODE_UNREACHABLE` line afterwards anyway for clarity.

**Review by ioannisg — APPROVED** [review]
